### ISSUE TITLE
Are these ES bugs or just wrong implementations by me?

### ISSUE THREAD

**Issue by apanimesh061** [elastic/elasticsearch#12819]
I have set the maximum shingle size as 3 for my index. When I am done with indexing I find that the shingles : 
1. `yield`
2. `_ yield`
3. `_ _ yield`

have the same start and ending offsets. So, if I would need the original sub-string I will always get `yield`!! Why is that and can it be corrected?

Another thing that I noticed is that the offsets of the shingles like:
1. `_ yield _`
2. `yield _`
3. `yield _ _`
have length such that the string has an extra space that the end. e.g. for `_ yield _` if there is a string `"not yield until"` then the offsets, is used on the original text, return `"yield until "` (notice trailing space). I actually would want `"not yield until"`.

Finally, is there a way we can avoid stop words while creating shingles? In my application the stop words don't play any role.

Following is the settings that I am using for the index:

    {
       "test_index": {
          "settings": {
             "index": {
                "refresh_interval": "60s",
                "number_of_shards": "5",
                "store": {
                   "type": "default"
                },
                "creation_date": "1439324341816",
                "analysis": {
                   "filter": {
                      "snowball_stop_words_en_EN": {
                         "type": "stop",
                         "stopwords_path": "snowball.stop"
                      },
                      "smart_stop_words_en_EN": {
                         "type": "stop",
                         "stopwords_path": "smart.stop"
                      },
                      "porter_stemmer_en_EN": {
                         "name": "porter",
                         "type": "stemmer"
                      },
                      "word_delimiter_en_EN": {
                         "type": "word_delimiter",
                         "stem_english_possessive": "true"
                      },
                      "default_stop_name_en_EN": {
                         "name": "_english_",
                         "type": "stop"
                      },
                      "preserve_original_en_EN": {
                         "type": "word_delimiter",
                         "preserve_original": "true"
                      },
                      "apos_replace_en_EN": {
                         "pattern": ".*\\'$",
                         "type": "pattern_replace",
                         "replacement": ""
                      },
                      "shingle_filter_en_EN": {
                         "max_shingle_size": "3",
                         "min_shingle_size": "2",
                         "type": "shingle",
                         "output_unigrams": "true"
                      }
                   },
                   "analyzer": {
                      "test_analyzer": {
                         "filter": [
                            "lowercase",
                            "smart_stop_words_en_EN",
                            "preserve_original_en_EN",
                            "porter_stemmer_en_EN",
                            "asciifolding",
                            "apos_replace_en_EN",
                            "shingle_filter_en_EN"
                         ],
                         "type": "custom",
                         "tokenizer": "standard"
                      }
                   }
                },
                "number_of_replicas": "1",
                "version": {
                   "created": "1060099"
                },
                "uuid": "urrFTCEoThyuPjjBYCBgYQ"
             }
          }
       }
    }

I use the `test_analyzer` on the documents while indexing.

I am using ElasticSearch 1.6.0.

As an example I have even attached [link](https://www.dropbox.com/s/dipb0quha3wke1f/fft.txt?dl=0) to a text file that I was using.

Thanks.
Any suggestions are welcome.

**Comment by colings86** [comment]
This question has already been posted on the forums here: https://discuss.elastic.co/t/are-these-es-bugs-or-just-wrong-implementations-by-me/27228

Please keep the discussion in one place as it will avoid the conversation getting split between two threads which makes it hard to follow both for current participants and for people looking for the same information in the future.

**Comment by clintongormley** [comment]
@colings86 actually i think this is worth discussing here.  there may be things we can improve.  Reopening

**Comment by clintongormley** [comment]
@apanimesh061 As a start, try getting rid of the word delimiter filter (first time i've seen it used twice???) but it is known to mess up the analysis chain.  Don't try to put every analysis variation into a single field - you just end up with garbage.  Instead break it down into a few fields used for different purposes, eg one with standard analyzer, one with stopwords and stemmer, etc

I'd be interested in seeing a full JSON replication (without word delimiter), ie something that I can run easily without having to build my own example :)

**Comment by apanimesh061** [comment]
@clintongormley I am pretty sure issue [#4307](https://github.com/elastic/elasticsearch/issues/4307#issuecomment-130239252) is related to this.
I have updated the settings as asked:

```
{
   "test_index": {
      "settings": {
         "index": {
            "refresh_interval": "60s",
            "number_of_shards": "5",
            "store": {
               "type": "default"
            },
            "creation_date": "1439381862716",
            "analysis": {
               "filter": {
                  "default_stop_name_en_EN": {
                     "name": "_english_",
                     "type": "stop"
                  },
                  "snowball_stop_words_en_EN": {
                     "type": "stop",
                     "stopwords_path": "snowball.stop"
                  },
                  "smart_stop_words_en_EN": {
                     "type": "stop",
                     "stopwords_path": "smart.stop"
                  },
                  "apos_replace_en_EN": {
                     "pattern": ".*\\'$",
                     "type": "pattern_replace",
                     "replacement": ""
                  },
                  "shingle_filter_en_EN": {
                     "max_shingle_size": "3",
                     "min_shingle_size": "2",
                     "type": "shingle",
                     "output_unigrams": "true"
                  },
                  "porter_stemmer_en_EN": {
                     "name": "porter",
                     "type": "stemmer"
                  }
               },
               "analyzer": {
                  "test_analyzer": {
                     "type": "custom",
                     "filter": [
                        "lowercase",
                        "smart_stop_words_en_EN",
                        "porter_stemmer_en_EN",
                        "asciifolding",
                        "apos_replace_en_EN",
                        "shingle_filter_en_EN"
                     ],
                     "tokenizer": "standard"
                  }
               }
            },
            "number_of_replicas": "1",
            "version": {
               "created": "1060099"
            },
            "uuid": "uvmOqkayTOKy8H3FpFH6zA"
         }
      }
   }
}
```

**Comment by clintongormley** [comment]
@apanimesh061 when I said "full JSON replication" I meant: a series of curl commands that i can copy and paste, along with "this request produces this output, while i think it should be this instead".  otherwise I have to spend time trying to guess which parts you think are wrong.

**Comment by clintongormley** [comment]
No more feedback. Closing
